### PR TITLE
#1387: Add `TextDirectoryCorpus` that yields one doc per file recursi…

### LINKED ISSUE #1387: Add `TextDirectoryCorpus` that yields one doc per file recursively read from directory
#### Description
Plain text corpuses are sometimes represented using a directory structure with, for instance a top-level directory and subdirectories that represent categories. Then within each subdirectory, each file might be a document. The nesting may run deeper in such directory structures to reflect additional categorization. The popular 20 newsgroups dataset has such a structure. Here is a subset of that structure:
```
20news-18828/
|-- alt.atheism
|   |-- 49960
|   |-- 51060
|   |-- 51119
|-- comp.graphics
|   |-- 37261
|   |-- 37913
|   |-- 37914
|   |-- 37915
|   |-- 37916
|   |-- 37917
|   |-- 37918
|-- comp.os.ms-windows.misc
|   |-- 10000
|   |-- 10001
|   |-- 10002
|   |-- 10003
|   |-- 10004
|   |-- 10005 
```

It would be useful to have a gensim corpus available to handle this sort of corpus structure.

#### Steps/Code/Corpus to Reproduce

I'm envisioning something like this:
```
class TextDirectoryCorpus(TextCorpus):
    """Read documents recursively from a directory,
    where each file is interpreted as a plain text document.
    """

    def __init__(self, input, metadata=False, min_depth=0, max_depth=None, pattern=None,
                 exclude_pattern=None, **kwargs):
        self.min_depth = min_depth
        self.max_depth = sys.maxsize if max_depth is None else max_depth
        self.pattern = pattern
        self.exclude_pattern = exclude_pattern
        super(TextDirectoryCorpus, self).__init__(input, metadata, **kwargs)
    ....

corpus = textcorpus.TextDirectoryCorpus('20_newsgroups', exclude_pattern="README.*")
paths = corpus.iter_filepaths()
texts = corpus.get_texts()
```

#### Versions
This should be available in all versions.

### PR DESCRIPTION
**Pull request by macks22**
…vely read from directory. Resolves #1387.

### CODE REVIEW

**Comment by macks22** [comment]
@piskvorky I know we had discussed refactoring the TextCorpus variants in general; do you think it's best to merge this and then open up another issue for the more general architectural changes we were discussing, or should I add that sort of stuff to this PR? Thanks!

**Comment by macks22** [comment]
@menshikh-iv do you believe this requires any changes? Also, any thoughts on the question I posed to @piskvorky?

File: gensim/corpora/textcorpus.py [hunk 1]
```diff
@@ -49,35 +65,104 @@ class TextCorpus(interfaces.CorpusABC):
...
+    1.  lowercase and convert to unicode; assumes utf8 encoding
+    2.  deaccent (asciifolding)
+    3.  collapse multiple whitespaces into a single one
+    4.  tokenize by splitting on whitespace
+    5.  remove words less than 3 characters long
+    6.  remove stopwords; see `gensim.parsing.preprocessing` for the list of stopwords
 
     """
-    def __init__(self, input=None):
+    def __init__(self, input=None, metadata=False, character_filters=None, tokenizer=None,
```
Comment (menshikh-iv) [inline gensim/corpora/textcorpus.py]: Please, add documentation for a new parameters in [google-docstring format](https://github.com/RaRe-Technologies/gensim/wiki/Developer-page#docstrings)

File: gensim/corpora/textcorpus.py [hunk 2]
```diff
@@ -49,35 +65,104 @@ class TextCorpus(interfaces.CorpusABC):
...
         if input is not None:
             self.dictionary.add_documents(self.get_texts())
         else:
             logger.warning("No input document stream provided; assuming "
                            "dictionary will be initialized some other way.")
 
     def __iter__(self):
-        """
-        The function that defines a corpus.
+        """The function that defines a corpus.
```
Comment (menshikh-iv) [inline gensim/corpora/textcorpus.py]: Please use google-docstring format everywhere

File: gensim/corpora/textcorpus.py [hunk 3]
```diff
@@ -125,5 +211,121 @@ def __len__(self):
...
 # endclass TextCorpus
+
+
+class TextDirectoryCorpus(TextCorpus):
+    """Read documents recursively from a directory,
+    where each file is interpreted as a plain text document.
+    """
+
+    def __init__(self, input, metadata=False, min_depth=0, max_depth=None, pattern=None,
+                 exclude_pattern=None, **kwargs):
```
Comment (menshikh-iv) [inline gensim/corpora/textcorpus.py]: Add docstring with parameter description

File: gensim/corpora/textcorpus.py [hunk 4]
```diff
@@ -125,5 +211,121 @@ def __len__(self):
             # cache the corpus length
             self.length = sum(1 for _ in self.get_texts())
         return self.length
-
 # endclass TextCorpus
+
+
+class TextDirectoryCorpus(TextCorpus):
+    """Read documents recursively from a directory,
+    where each file is interpreted as a plain text document.
```
Comment (menshikh-iv) [inline gensim/corpora/textcorpus.py]: Maybe iterate by line (not by file), what do you think about it?

**Comment by menshikh-iv** [comment]
Can you add short example of usage of this feature with non-trivial tree (as "integration" test)
Example of tree:
```
.
├── a_folder/
│        ├── 0.txt
├── b_folder/
│       ├── 1.txt
│       ├── 2.txt
│       ├── c_folder/
│               ├── 3.txt
```

File: gensim/corpora/textcorpus.py (hunk 4 above)
Comment (macks22) [inline gensim/corpora/textcorpus.py]: I think we could add an option for that, but the corpora this was designed for (20 newsgroups, Wiki-Movie from coherence papers) are one document per file.

File: gensim/corpora/textcorpus.py (hunk 2 above)
Comment (macks22) [inline gensim/corpora/textcorpus.py]: done

File: gensim/corpora/textcorpus.py (hunk 4 above)
Comment (macks22) [inline gensim/corpora/textcorpus.py]: added a new argument `lines_are_documents` along with a test that shows usage

File: gensim/corpora/textcorpus.py (hunk 3 above)
Comment (macks22) [inline gensim/corpora/textcorpus.py]: done

File: gensim/corpora/textcorpus.py (hunk 1 above)
Comment (macks22) [inline gensim/corpora/textcorpus.py]: done

**Comment by macks22** [comment]
@menshikh-iv I've addressed your PR comments; thank you for the review!

Sorry about the big blob of commits since your review -- I had to rebase and resolve several conflicts with gensim upstream. The first commit, d41b2c4, contains my changes in response to your review.

**Comment by macks22** [comment]
@menshikh-iv @piskvorky Looks like I botched up this PR with the rebase I did to resolve conflicts recently. I've opened up PR #1459 to replace it.

**Pull request closed without merging**